### PR TITLE
 fallback support FacotryBean

### PR DESCRIPTION
**Pull request by min1854**
fallback is FactoryBean IllegalStateException

https://github.com/spring-cloud/spring-cloud-openfeign/issues/822

### CODE REVIEW

**Review by OlgaMaciaszek — CHANGES_REQUESTED** [review]
Hello @min1854, thanks for submitting the PR.  Please add your name and surname with the `@author` tag in the Javadoc of any class you've modified. Please add tests.

**Comment by OlgaMaciaszek** [comment]
Fixes gh-822.

**Comment by min1854** [comment]
> Hello @min1854, thanks for submitting the PR. Please add your name and surname with the `@author` tag in the Javadoc of any class you've modified. Please add tests.

Ok, thank you. 
My localhost git user.name name is different from the github account name, do I need to change it to the github account name

**Comment by OlgaMaciaszek** [comment]
> > Hello @min1854, thanks for submitting the PR. Please add your name and surname with the `@author` tag in the Javadoc of any class you've modified. Please add tests.
> 
> Ok, thank you. My localhost git user.name name is different from the github account name, do I need to change it to the github account name

No, just put the correct one in the javadoc.

**Comment by min1854** [comment]
> > > Hello @min1854, thanks for submitting the PR. Please add your name and surname with the `@author` tag in the Javadoc of any class you've modified. Please add tests.
> > 
> > 
> > Ok, thank you. My localhost git user.name name is different from the github account name, do I need to change it to the github account name
> 
> No, just put the correct one in the javadoc.



> > > Hello @min1854, thanks for submitting the PR. Please add your name and surname with the `@author` tag in the Javadoc of any class you've modified. Please add tests.
> > 
> > 
> > Ok, thank you. My localhost git user.name name is different from the github account name, do I need to change it to the github account name
> 
> No, just put the correct one in the javadoc.

I would like to ask if my javadoc @author  should use local user.name or should use the github account name

**Comment by spencergibb** [comment]
It should be your name, not a user or github account name

**Comment by min1854** [comment]
> It should be your name, not a user or github account name

Ok, thanks.

File: spring-cloud-openfeign-core/src/test/java/org/springframework/cloud/openfeign/circuitbreaker/FallbackSupportFactoryBeanTests.java [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+	@FeignClient(name = "original", url = "https://original", fallback = OriginalFeignFallback.class)
+	public interface OriginalFeign {
+
+		@GetMapping("/")
+		String get();
+
+	}
+
+	@FeignClient(name = "factoryBean", url = "https://factoryBean", fallback = FactoryBeanFallbackFeignFallback.class)
+	public interface FactoryBeanFallbackFeign {
```
Comment (OlgaMaciaszek) [inline spring-cloud-openfeign-core/src/test/java/org/springframework/cloud/openfeign/circuitbreaker/FallbackSupportFactoryBeanTests.java:106]: Please remove the unnecessary `public` keyword.

File: spring-cloud-openfeign-core/src/test/java/org/springframework/cloud/openfeign/circuitbreaker/FallbackSupportFactoryBeanTests.java [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+	@Configuration(proxyBeanMethods = false)
+	@EnableFeignClients(clients = { FallbackSupportFactoryBeanTests.OriginalFeign.class,
+			FallbackSupportFactoryBeanTests.FactoryBeanFallbackFeign.class })
+	@EnableAutoConfiguration
+	public static class TestConfiguration {
+
+	}
+
+	@FeignClient(name = "original", url = "https://original", fallback = OriginalFeignFallback.class)
+	public interface OriginalFeign {
```
Comment (OlgaMaciaszek) [inline spring-cloud-openfeign-core/src/test/java/org/springframework/cloud/openfeign/circuitbreaker/FallbackSupportFactoryBeanTests.java:98]: Please remove the unnecessary `public` keyword.

File: spring-cloud-openfeign-core/src/test/java/org/springframework/cloud/openfeign/circuitbreaker/FallbackSupportFactoryBeanTests.java [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+				}
+
+				@Override
+				public void configureDefault(Function defaultConfiguration) {
+
+				}
+			}).withPropertyValues("spring.cloud.openfeign.circuitbreaker.enabled=true");
+
+	@Test
+	public void factoryBeanFallback() {
```
Comment (OlgaMaciaszek) [inline spring-cloud-openfeign-core/src/test/java/org/springframework/cloud/openfeign/circuitbreaker/FallbackSupportFactoryBeanTests.java:79]: Please rename the test to `shouldRunFallbackFromBeanOrFactoryBean`.

**Review by OlgaMaciaszek — CHANGES_REQUESTED** [review]
Thanks for adding the test @min1854. Have added some comments, practically all cosmetic. Please address them, and we'll be able to get it merged.

File: spring-cloud-openfeign-core/src/test/java/org/springframework/cloud/openfeign/circuitbreaker/FallbackSupportFactoryBeanTests.java [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+import org.springframework.cloud.openfeign.FeignClient;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.web.bind.annotation.GetMapping;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+/**
+ * @author 黄学敏（huangxuemin)
+ */
+public class FallbackSupportFactoryBeanTests {
```
[the remark this answers is not in the document]
Comment (min1854) [reply spring-cloud-openfeign-core/src/test/java/org/springframework/cloud/openfeign/circuitbreaker/FallbackSupportFactoryBeanTests.java:42]: class modifier change to private?

[the remark this answers is not in the document]
Comment (min1854) [reply spring-cloud-openfeign-core/src/test/java/org/springframework/cloud/openfeign/circuitbreaker/FallbackSupportFactoryBeanTests.java:42]: image
Or is that it?

[the remark this answers is not in the document]
Comment (OlgaMaciaszek) [reply spring-cloud-openfeign-core/src/test/java/org/springframework/cloud/openfeign/circuitbreaker/FallbackSupportFactoryBeanTests.java:42]: No, just package level: `class FallbackSupportFactoryBeanTests`, with no added modifier.

File: spring-cloud-openfeign-core/src/test/java/org/springframework/cloud/openfeign/circuitbreaker/FallbackSupportFactoryBeanTests.java [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+
+	@Test
+	public void factoryBeanFallback() {
+
+		runner.run(ctx -> {
+			assertThat(ctx.getBean(OriginalFeign.class).get().equals(ORIGINAL_FALLBACK_MESSAGE)).isTrue();
+			assertThat(ctx.getBean(FactoryBeanFallbackFeign.class).get().equals(FACTORY_BEAN_FALLBACK_MESSAGE))
+					.isTrue();
+		});
+
```
[the remark this answers is not in the document]
Comment (min1854) [reply spring-cloud-openfeign-core/src/test/java/org/springframework/cloud/openfeign/circuitbreaker/FallbackSupportFactoryBeanTests.java:86]: Are these the empty lines

image

[the remark this answers is not in the document]
Comment (OlgaMaciaszek) [reply spring-cloud-openfeign-core/src/test/java/org/springframework/cloud/openfeign/circuitbreaker/FallbackSupportFactoryBeanTests.java:86]: Yes, those were the lines I meant.

[the remark this answers is not in the document]
Comment (min1854) [reply spring-cloud-openfeign-core/src/test/java/org/springframework/cloud/openfeign/circuitbreaker/FallbackSupportFactoryBeanTests.java:86]: final result 
image

Is that OK?

[the remark this answers is not in the document]
Comment (OlgaMaciaszek) [reply spring-cloud-openfeign-core/src/test/java/org/springframework/cloud/openfeign/circuitbreaker/FallbackSupportFactoryBeanTests.java:86]: Yes, it is @min1854. Please also address the other comments. There's no need to add screenshot - just address all the comments and push the change - will review again once it's all done.

[the remark this answers is not in the document]
Comment (min1854) [reply spring-cloud-openfeign-core/src/test/java/org/springframework/cloud/openfeign/circuitbreaker/FallbackSupportFactoryBeanTests.java]: Ok, I really appreciate your help

[the remark this answers is not in the document]
Comment (OlgaMaciaszek) [reply spring-cloud-openfeign-core/src/test/java/org/springframework/cloud/openfeign/circuitbreaker/FallbackSupportFactoryBeanTests.java]: Thanks for working on the issue :)

**Comment by min1854** [comment]
> Thanks for adding the test @min1854. Have added some comments, practically all cosmetic. Please address them, and we'll be able to get it merged.

Am I done?Is there anything else I need to do?

**Comment by min1854** [comment]
> Thanks for adding the test @min1854. Have added some comments, practically all cosmetic. Please address them, and we'll be able to get it merged.

Should I not click on 'conversation as resolved.'.Is that right. I'm sorry. I think I misunderstood

File: spring-cloud-openfeign-core/src/test/java/org/springframework/cloud/openfeign/circuitbreaker/FallbackSupportFactoryBeanTests.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (OlgaMaciaszek) [reply spring-cloud-openfeign-core/src/test/java/org/springframework/cloud/openfeign/circuitbreaker/FallbackSupportFactoryBeanTests.java:42]: @min1854 do not mark them as resolved. We'll resolve them. Yes, it's like the code from screenshot. Please commit and. push the changed code.

File: spring-cloud-openfeign-core/src/test/java/org/springframework/cloud/openfeign/circuitbreaker/FallbackSupportFactoryBeanTests.java [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+				}
+
+				@Override
+				public void configureDefault(Function defaultConfiguration) {
+
+				}
+			}).withPropertyValues("spring.cloud.openfeign.circuitbreaker.enabled=true");
+
+	@Test
+	public void shouldRunFallbackFromBeanOrFactoryBean() {
```
Comment (OlgaMaciaszek) [inline spring-cloud-openfeign-core/src/test/java/org/springframework/cloud/openfeign/circuitbreaker/FallbackSupportFactoryBeanTests.java:79]: Please remove the unnecessary `public` keyword here as well.

**Review by OlgaMaciaszek — CHANGES_REQUESTED** [review]
Hello, @min1854, there are still unaddressed comments. Please add changes.

File: spring-cloud-openfeign-core/src/test/java/org/springframework/cloud/openfeign/circuitbreaker/FallbackSupportFactoryBeanTests.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (OlgaMaciaszek) [reply spring-cloud-openfeign-core/src/test/java/org/springframework/cloud/openfeign/circuitbreaker/FallbackSupportFactoryBeanTests.java:42]: @min1854 it still is `public class FallbackSupportFactoryBeanTests `. Please change to `class FallbackSupportFactoryBeanTests`, without the `public` keyword.

File: spring-cloud-openfeign-core/src/test/java/org/springframework/cloud/openfeign/circuitbreaker/FallbackSupportFactoryBeanTests.java [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+			assertThat(ctx.getBean(FactoryBeanFallbackFeign.class).get().equals(FACTORY_BEAN_FALLBACK_MESSAGE))
+					.isTrue();
+		});
+	}
+
+	@Configuration(proxyBeanMethods = false)
+	@EnableFeignClients(clients = { FallbackSupportFactoryBeanTests.OriginalFeign.class,
+			FallbackSupportFactoryBeanTests.FactoryBeanFallbackFeign.class })
+	@EnableAutoConfiguration
+	public static class TestConfiguration {
```
Comment (OlgaMaciaszek) [inline spring-cloud-openfeign-core/src/test/java/org/springframework/cloud/openfeign/circuitbreaker/FallbackSupportFactoryBeanTests.java:91]: Could this be changed to a `private static class` as well.

**Review by OlgaMaciaszek — CHANGES_REQUESTED** [review]
Thanks, @min1854, have added 2 more comments, including one in a previous conversation. These are just tiny cosmetic changes. Please implement them and we'll be able to get this merged.

**Review by OlgaMaciaszek — APPROVED** [review]
Thanks, @min1854. LGTM.

**Comment by OlgaMaciaszek** [comment]
No, it's fine. This issue is now closed. Thanks for working on it.

**Comment by min1854** [comment]
Ok, thank you

**Pull request merged**